src/null-safety/index.md

```diff
@@ -1,39 +1,247 @@
+
+As an example, look at the following code. What do you think the value and type of `uhOh` are?
+
+```dart
+var myMap = <String, int>{'one': 1};
+var uhOh = myMap['two'];
+```
+
+The answer is that `uhOh` is null and has type `int?`. 
+
+Like lists and sets, maps can have a variety of type permutations.
+
+|----------------------+-------------------------------+-------------------------|
+|Type                  | Can an item (int)<br>be null? | Can the map<br>be null? |
+|----------------------|-------------------------------|-------------------------|
+| `Map<String, int>`   | No*                           | No                      |
+| `Map<String, int>?`  | No*                           | {{yes}}                 |
+| `Map<String, int?>`  | {{yes}}                       | No                      |
+| `Map<String, int?>?` | {{yes}}                       | {{yes}}                 |
+{:.table .table-striped}
+
+_\* Even when all the int values in the map are non-null,
+when you use an invalid key to do a map lookup, the returned value is null._
+
+Because map lookups can return null,
+you can't assign them to non-nullable variables:
+
+```dart
+// Assigning a lookup result to a non-nullable
+// variable causes an error.
+int value = <String, int>{'one': 1}['one']; // ERROR
+```
+
+**[QUESTION: Will the analyzer ever be able to detect that is non-null?]**
+
+One workaround is to change the type of the variable to be nullable:
+
+```dart
+int? value = <String, int>{'one': 1}['one']; // OK
+```
+
+Another way to fix the problem —
+if you're sure the lookup succeeds —
+is to add a `!`:
+
+```dart
+int value = <String, int>{'one': 1}['one']!; // OK
+```
+
+A safer approach is to use the lookup value only if it's not null.
+You can test its value using
+an [`if` statement][] or the [`??` operator][].
+Here's an example of using the value `0` if the lookup returns a null value:
+
+```dart
+var aList = <String, int>{'one': 1};
+...
+int value = aList['one'] ?? 0;
+```
+
+[`if` statement]: /guides/language/language-tour#if-and-else
+[`??` operator]: /guides/language/language-tour#conditional-expressions
+
+
+## Fixing analysis errors
+
+With sound null safety, the Dart analyzer generates errors when
+it finds a nullable value where a non-null value is required.
+That's not as bad as it sounds:
+the analyzer can often recognize when
+a variable or expression inside a function has
+a nullable type but can't have a null value.
+However, the analyzer can't model the flow of your whole application,
+so it can't predict the value of a global variable or class field.
+
+If you know a value can't be null
+but it has a nullable type,
+consider using `!`, `late`, or `as`.
+Examples of using `!` and `late` are in the [first section of this page][].
+Here's an example of using `as` to convert an `int?` to a `int`:
+
+```dart
+return maybeInt(3) as int;
```

Comment:
Don't do (or recommend) using `as` like that. Always use `!` 
... unless you actually want to check for a different type, like `maybeNum() as int`.